.release-notes/4567.md

```diff
@@ -0,0 +1,3 @@
+## Deterministic cerealisation for cross binary communication
+
+Pony built-in serialisation can now be used between binaries compiled with the same version of the pony compiler (this would likely result in segfaults previously). Cross binary serialisation will only work for binaries of the same bit width (32 bit vs 64 bit), data model (ilp32, lp64, or llp64), and endianness (big endian or little endian) but is not limited to a single platform (for example: one can mix and match x86_64 linux and aarch64 linux because they have the same bitwidth, data model, and endianness).
```

Comment:
I think we need an additional section of "serialization can lead to bad things", "this can be an attack model on pony programs and should only be used with trusted input", and what not.

Reply:
hmmm.. the `serialise` package currently has the following (some/all of which can be copied to the release notes and also be updated for the changes in this PR):

> Deserialisation is fundamentally unsafe currently: there isn't yet a
verification pass to check that the resulting object graph maintains a
well-formed heap or that individual objects maintain any expected local
invariants. However, if only "trusted" data (i.e. data produced by Pony
serialisation from the same binary) is deserialised, it will always maintain a
well-formed heap and all object invariants.
>
> Note that serialised data is not usable between different Pony binaries. This is
due to the use of type identifiers rather than a heavy-weight self-describing
serialisation schema. This also means it isn't safe to deserialise something
serialised by the same program compiled for a different platform.
>
> The [Serialise.signature](serialise-Serialise.md#signature) method is provided
for the purposes of comparing communicating Pony binaries to determine if they
are the same. Confirming this before deserialising data can help mitigate the
risk of accidental serialisation across different Pony binaries, but does not on
its own address the security issues of accepting data from untrusted sources.
------
a few notes to ensure everyone understands the full scope/limitations of this PR:

1. > The [Serialise.signature](serialise-Serialise.md#signature) method is provided
for the purposes of comparing communicating Pony binaries to determine if they
are the same. 

    i missed this in the changes so far and this method needs to be updated to return the signature of the target pony runtime rather than the program..
2. > However, if only "trusted" data (i.e. data produced by Pony
serialisation from the same binary) is deserialised, it will always maintain a
well-formed heap and all object invariants.

    * with this PR, we can only `maintain a well-formed heap and all object invariants` if the binary doing the `deserialisation` knows/uses all the types that were used for `serialisation` or else the program will assert/fail with `ponyint_assert_fail("deserialise offset invalid", __FILE__, __LINE__, __func__)` during `deserialisation` (`deserialise offset invalid` probably needs rewording to be more explicit about the issue/error)..

    * `maintain a well-formed heap and all object invariants` cannot be guaranteed for all combinations of programs because the `serialise_id` is only based on the `type name` and not the type's `name and fields/memory layout` (or `full ast`).. unfortunately, this will cause data corruption (and possibly memory clobbering) and not an assert/fail like the previous bullet because the `serialise_id` will be the same between the two binaries even though it shouldn't be (because a type's field layout changed) due to the limitation on how `serialise_id` is determined currently..


given the above, especially the second bullet under `2`, maybe this PR should be reframed to be `internal changes in support of cross binary serialisation` until the `serialise_id` can be based on either the `full ast` or the `name and fields/memory layout` for each type to remove that caveat/concern (if so, it would make sense to defer updating `Serialise.signature` until that time also)?

Reply:
I think noting at a high-level a couple of issues and referring people to the docs on stdlib.ponylang.io for the full list of reasons to not use serialisation is good.

Reply:
> So basically, this doesn't get a changelog entry for now and we don't tell people about the change. that is my interpretation of "internal". is that yours?

yes, unless we're ok with the large footgun that is `the second bullet under "2"`..

Reply:
yup. i think internal is a good idea. i'll remove the label. you can remove the release notes if you want, otherwise the bot will toss them on merge.

Reply:
gonna leave it.. gotta keep the bots employed or they might revolt...

Reply:
@dipinhora we discussed this during sync. can you update the serialization package documentation to note the changes that were documented in the release notes here?

Reply:
and remove the release notes.

Reply:
sure.. but technically:

> Pony built-in serialisation can now be used between binaries compiled with the same version of the pony compiler (this would likely result in segfaults previously). Cross binary serialisation will only work for binaries of the same bit width (32 bit vs 64 bit), data model (ilp32, lp64, or llp64), and endianness (big endian or little endian) but is not limited to a single platform (for example: one can mix and match x86_64 linux and aarch64 linux because they have the same bitwidth, data model, and endianness).

isn't true because the [Serialise.signature](https://github.com/ponylang/ponyc/pull/serialise-Serialise.md#signature) function still generates a signature that is unique to each program and the `serialise` package recommends using it as a safeguard to ensure that serialisation will be safe to use..

Reply:
serialise package docs updated and release notes removed..

note: the serialise package docs are technically no longer correct as per my last comment (https://github.com/ponylang/ponyc/pull/4567#discussion_r1889128715) because the [Serialise.signature](https://github.com/ponylang/ponyc/pull/serialise-Serialise.md#signature) function still generates a signature that is unique to each program and the serialise package recommends using it as a safeguard to ensure that serialisation will be safe to use and this will prevent cross binary serialisation from working if that recommendation is followed because the signatures will not match..